codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -211,134 +263,245 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+            frame.release();
             throw new IllegalArgumentException("Last stream id must not be set on GOAWAY frame");
         }
 
-        int lastStreamCreated = http2Handler.connection().remote().lastStreamCreated();
-        int lastStreamId = lastStreamCreated + frame.extraStreamIds() * 2;
+        int lastStreamCreated = connection().remote().lastStreamCreated();
+        long lastStreamId = lastStreamCreated + ((long) frame.extraStreamIds()) * 2;
         // Check if the computation overflowed.
-        if (lastStreamId < lastStreamCreated) {
+        if (lastStreamId > Integer.MAX_VALUE) {
             lastStreamId = Integer.MAX_VALUE;
         }
-        http2Handler.goAway(
-                http2HandlerCtx, lastStreamId, frame.errorCode(), frame.content().retain(), promise);
+        goAway(ctx, (int) lastStreamId, frame.errorCode(), frame.content(), promise);
     }
 
-    private void writeStreamFrame(Http2StreamFrame frame, ChannelPromise promise) {
-        if (frame instanceof Http2DataFrame) {
-            Http2DataFrame dataFrame = (Http2DataFrame) frame;
-            http2Handler.encoder().writeData(http2HandlerCtx, frame.streamId(), dataFrame.content().retain(),
-                                             dataFrame.padding(), dataFrame.isEndStream(), promise);
-        } else if (frame instanceof Http2HeadersFrame) {
-            writeHeadersFrame((Http2HeadersFrame) frame, promise);
-        } else if (frame instanceof Http2ResetFrame) {
-            Http2ResetFrame rstFrame = (Http2ResetFrame) frame;
-            http2Handler.resetStream(http2HandlerCtx, frame.streamId(), rstFrame.errorCode(), promise);
+    private void writeHeadersFrame(
+            final ChannelHandlerContext ctx, Http2HeadersFrame headersFrame, final ChannelPromise promise) {
+
+        if (isStreamIdValid(headersFrame.stream().id())) {
+            encoder().writeHeaders(ctx, headersFrame.stream().id(), headersFrame.headers(), headersFrame.padding(),
+                    headersFrame.isEndStream(), promise);
         } else {
-            throw new UnsupportedMessageTypeException(frame);
+            final DefaultHttp2FrameStream stream = (DefaultHttp2FrameStream) headersFrame.stream();
+            final Http2Connection connection = connection();
+            final int streamId = connection.local().incrementAndGetNextStreamId();
+            if (streamId < 0) {
+                promise.setFailure(new Http2NoMoreStreamIdsException());
+                return;
+            }
+            stream.id = streamId;
+
+            // TODO: This depends on the fact that we currently add streams in a synchron manner. We should investigate
+            // how to refactor this later on when we consolidate some layers.
+            assert frameStreamToInitialize == null;
+            frameStreamToInitialize = stream;
+
+            // TODO(buchgr): Once Http2Stream2 and Http2Stream are merged this is no longer necessary.
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)